src/lib/helpers.ts

```diff
@@ -145,3 +145,45 @@ export function decodeJWTPayload(token: string) {
   const base64Url = parts[1]
   return JSON.parse(decodeBase64URL(base64Url))
 }
+
+/**
+ * Creates a promise that resolves to null after some time.
+ */
+export function sleep(time: number): Promise<null> {
+  return new Promise((accept) => {
+    setTimeout(() => accept(null), time)
+  })
+}
+
+/**
+ * Converts the provided async function into a retryable function. Each result
+ * or thrown error is sent to the isRetryable function which should return true
+ * if the function should run again.
+ */
+export function retryable<T>(
+  fn: (attempt: number) => Promise<T>,
+  isRetryable: (attempt: number, error: any | null, result?: T) => boolean
+): Promise<T> {
+  const promise = new Promise<T>((accept, reject) => {
+    // eslint-disable-next-line @typescript-eslint/no-extra-semi
+    ;(async () => {
+      for (let attempt = 0; attempt < Infinity; attempt++) {
+        try {
+          const result = await fn(attempt)
```

Comment:
should we call the `sleep` function here instead of inside the `fn` ? i think it should be pretty standard that any request that needs to be retried should follow an exponential backoff

Reply:
I'd rather not call any sleeping because this is a helper function designed to easily build retry logic. Not all retries need to have exponential backoff (for example if there's no internet on the device). By keeping this one small we can be certain that it will always behave properly.